industrial_taxonomy/pipeline/glass_house/flow.py

```diff
@@ -174,23 +168,41 @@ def end(self):
 
         self.full_top_matches = (
             self.top_matches.merge(
-                self.names_y.reset_index().rename(columns={"index": "index_y"}),
+                self.glass_names.reset_index().rename(columns={"name": "org_name"}),
                 left_on="y",
                 right_index=True,
                 validate="1:1",
             )
             .merge(
-                self.names_x.reset_index().rename(columns={"index": "index_x"}),
+                self.ch_names.reset_index().rename(columns={"name": "company_name"}),
                 left_on="x",
                 right_index=True,
                 validate="m:1",
             )
-            .rename(columns={"name_x": "company_name", "name_y": "org_name"})
             .drop(["x", "y"], axis=1)
             .to_dict("records")
         )
         del self.top_matches
 
+    def _persist_chunk(self, chunk_id: int, chunk: "DataFrame") -> int:
+        key = f"match-chunks-{chunk_id}"
+        with S3(run=self) as s3:
+            s3.put(key, chunk.to_csv())
+        print(f"Persisted chunk {chunk_id} of shape {chunk.shape}")
+        return chunk_id
+
+    def _get_chunk(self, chunk_id: int) -> "DataFrame":
+        from io import StringIO
+        from pandas import read_csv
+
+        key = f"match-chunks-{chunk_id}"
+        try:
+            with S3(run=self) as s3:
+                return read_csv(StringIO(s3.get(key).text))
+        except MetaflowS3NotFound:  # If flow resumed, could be under `origin_run_id`
+            with S3(run=Run(f"{current.flow_name}/{current.origin_run_id}")) as s3:
+                return read_csv(StringIO(s3.get(key).text))
```

Comment:
That identification of re-usable components and refactoring of those out to utils should be a phase at the end of a project and that this needs to be embedded into the project lifecycle.
(though of course people can optionally be more pro-active and do things earlier as long as it doesn't hold up the project)

Reply:
Though we should be careful to:
* Be selective about what we choose to have as re-usable components
  * Needs to be worth both the initial effort to develop into a re-usable component AND the maintenance burden
* Properly (re-)design anything chosen as a re-usable component to limit complexity and maximise composability (and hence the number of places it can be reused)

So many things will end up not being worth having as re-usable components but could be added to a "Cookbook" which can be a document/site/repo that links to the feature in the project's code and has an accompanying explanation about how the problem is tackled and why.